src/backend/distributed/commands/table.c

```diff
@@ -162,31 +163,61 @@ PreprocessDropTableStmt(Node *node, const char *queryString,
 
 
 /*
- * PostprocessCreateTableStmt takes CreateStmt object as a parameter and errors
- * out if it creates a table with a foreign key that references to a citus local
- * table.
+ * PostprocessCreateTableStmt takes CreateStmt object as a parameter and
+ * processes foreign keys on relation via PostprocessCreateTableStmtForeignKeys
+ * function.
  *
  * This function also processes CREATE TABLE ... PARTITION OF statements via
  * PostprocessCreateTableStmtPartitionOf function.
  */
 void
 PostprocessCreateTableStmt(CreateStmt *createStatement, const char *queryString)
+{
+	PostprocessCreateTableStmtForeignKeys(createStatement);
+
+	if (createStatement->inhRelations != NIL && createStatement->partbound != NULL)
+	{
+		/* process CREATE TABLE ... PARTITION OF command */
+		PostprocessCreateTableStmtPartitionOf(createStatement, queryString);
+	}
+}
+
+
+/*
+ * PostprocessCreateTableStmtForeignKeys drops ands re-defines foreign keys
+ * defined by given CREATE TABLE command if command defined any foreign to
+ * reference or citus local tables.
+ */
+static void
+PostprocessCreateTableStmtForeignKeys(CreateStmt *createStatement)
 {
 	/*
 	 * Relation must exist and it is already locked as standard process utility
 	 * is already executed.
 	 */
 	bool missingOk = false;
 	Oid relationId = RangeVarGetRelid(createStatement->relation, NoLock, missingOk);
-	if (HasForeignKeyToCitusLocalTable(relationId))
+
+	if (!(TableReferenced(relationId) || TableReferencing(relationId)))
 	{
-		ErrorOutForFKeyBetweenPostgresAndCitusLocalTable(relationId);
+		return;
 	}
 
-	if (createStatement->inhRelations != NIL && createStatement->partbound != NULL)
+	List *nonDistTableForeignKeyIdList =
+		GetForeignKeyOids(relationId, INCLUDE_REFERENCING_CONSTRAINTS |
+						  INCLUDE_CITUS_LOCAL_TABLES |
+						  INCLUDE_REFERENCE_TABLES);
+	bool hasForeignKeyToNonDistTable = list_length(nonDistTableForeignKeyIdList) != 0;
+	if (hasForeignKeyToNonDistTable)
 	{
-		/* process CREATE TABLE ... PARTITION OF command */
-		PostprocessCreateTableStmtPartitionOf(createStatement, queryString);
+		/*
+		 * Drop and re-define foreign keys so that our ALTER TABLE hook does
+		 * the necessary job.
+		 */
+		List *relationFKeyCreationCommands =
+			GetReferencingForeignConstaintCommands(relationId);
+		DropRelationForeignKeys(relationId);
+		ExecuteAndLogDDLCommandList(relationFKeyCreationCommands);
```

Comment:
Now we re-create the same foreign keys. It could be very problematic. I guess we should be skipping the validation. Can we do something similar to inside `ExecuteAndLogDDLCommandList ` for foreign keys?
```C
		AlterTableType alterTableType = command->subtype;

		if (alterTableType == AT_AddConstraint)
		{
			Constraint *constraint = (Constraint *) command->def;
			if (constraint->contype == CONSTR_FOREIGN)
			{
				/*
				 * We only support ALTER TABLE ADD CONSTRAINT ... FOREIGN KEY, if it is
				 * only subcommand of ALTER TABLE. It was already checked in
				 * ErrorIfUnsupportedAlterTableStmt.
				 */
				Assert(list_length(commandList) == 1);

				rightRelationId = RangeVarGetRelid(constraint->pktable, lockmode,
												   alterTableStatement->missing_ok);

				/*
				 * Foreign constraint validations will be done in workers. If we do not
				 * set this flag, PostgreSQL tries to do additional checking when we drop
				 * to standard_ProcessUtility. standard_ProcessUtility tries to open new
				 * connections to workers to verify foreign constraints while original
				 * transaction is in process, which causes deadlock.
				 */
				constraint->skip_validation = true;
			}
		}
```

Reply:
seems that we could actually use `ADD table_constraint [ NOT VALID ]` syntax when generating such foreign key creation commands
https://www.postgresql.org/docs/current/sql-altertable.html

Reply:
But we cannot let it stay as invalid forever. We have some functions that modifies VALID/NOT VALID of the constraints, but `skip_validation ` simpler?

Reply:
fwiw, we have some infrastructure in multi_executor.c for skipping validation

Reply:
I'll handle this separately, so resolving here